warn/scrapers/ga.py

```diff
@@ -11,111 +12,125 @@
+        url = f"{self.base_url}?geoArea={self.area}&year={self.year}&step=search"
+        super().__init__(url)
 
-    current_year = datetime.now().year
-    first_year = 2002  # first available year
+    def get_response(self, scraper: scrapelib.Scraper) -> requests.models.Response:
+        """Get the response from the source, then writes it to a cache.
 
-    years = list(range(first_year, current_year + 1))
-    years.reverse()
+        Args:
+            scraper (scrapelib.Scraper): The scraper to use.
 
-    # Loop through the years and scrape them one by one
-    output_rows = []
-    for i, year in enumerate(years):
-        # Concoct the URL
-        url = f"{base_url}?geoArea={area}&year={year}&step=search"
-        cache_key = f"{state_code}/{year}.html"
-
-        # Read from cache if available and not this year or the year before
-        if cache.exists(cache_key) and year < current_year - 1:
-            html = cache.read(cache_key)
-        else:
-            # Otherwise, go request it
-            page = utils.get_url(url)
-            html = page.text
-            cache.write(cache_key, html)
-
-        # Scrape out the table
-        new_rows = _parse_table(
-            html,
-            "emplrList",
-            include_headers=i == 0,  # After the first loop, we can skip the headers
-        )
-
-        # Concatenate the rows
-        output_rows.extend(new_rows)
+        Returns:
+            requests.models.Response: The response from the source.
+        """
+        response = scraper.get(self.url)
 
-    # Write out the results
-    data_path = data_dir / f"{state_code}.csv"
-    utils.write_rows_to_csv(data_path, output_rows)
+        # Write the response to the cache
+        cache_key = f"ga/{self.year}.html"
+        html = response.text
+        cache = Cache()
+        cache.write(cache_key, html)
 
-    # Return the path to the CSV
-    return data_path
+        return response
 
 
-def _parse_table(html, id, include_headers=True):
-    """
-    Parse HTML table with given ID.
+class YearList(ListPage):
+    """The list of years to scrape."""
 
-    Keyword arguments:
-    html -- the HTML to parse
-    id -- the ID of the table to parse
-    include_headers -- whether to include the headers in the output (default True)
+    source = NullSource()
 
-    Returns: a list of rows
-    """
-    # Parse out data table
-    soup = BeautifulSoup(html, "html.parser")
-    table_list = soup.find_all(id=id)  # output is list-type
+    first_year = 2002  # first available year
+
+    @property
+    def current_year(self):
+        """Get the current year."""
+        return datetime.now().year
+
+    @property
+    def years(self):
+        """Get the list of years to scrape."""
+        return reversed(list(range(self.first_year, self.current_year + 1)))
 
-    # We expect the first table to be there with our data
-    assert len(table_list) > 0
-    table = table_list[0]
+    def process_page(self):
```

Comment:
Little awkward that we're processing a "page" that doesn't exist, but this is unfortunately necessary because the years are populated on the root page with JS.